azuredevops/client_repository_commit.go

```diff
@@ -0,0 +1,142 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azuredevops
+
+import (
+	"context"
+	"fmt"
+	"github.com/fluxcd/go-git-providers/gitprovider"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v6/git"
+)
+
+// CommitClient implements the gitprovider.CommitClient interface.
+var _ gitprovider.CommitClient = &CommitClient{}
+
+// CommitClient operates on the commits for a specific repository.
+type CommitClient struct {
+	*clientContext
+	ref gitprovider.RepositoryRef
+}
+
+// ListPage lists all repository commits of the given page and page size.
+// ListPage returns all available repository commits
+// using multiple paginated requests if needed.
+
+func (c *CommitClient) ListPage(ctx context.Context, branch string, perPage, page int) ([]gitprovider.Commit, error) {
+	dks, err := c.listPage(ctx, branch, perPage, page)
+	if err != nil {
+		return nil, err
+	}
+	// Cast to the generic []gitprovider.Commit
+	commits := make([]gitprovider.Commit, 0, len(dks))
+	for _, dk := range dks {
+		commits = append(commits, dk)
+	}
+	return commits, nil
+}
+
+func (c *CommitClient) listPage(ctx context.Context, branch string, perPage, page int) ([]*commitType, error) {
+	repositoryId := c.ref.GetRepository()
+	projectId := c.ref.GetIdentity()
+	branchRetrieved := git.GitVersionDescriptor{
+		Version: &branch,
+	}
+	searchCriteria := git.GitQueryCommitsCriteria{
+		Top:         &page,
+		ItemVersion: &branchRetrieved,
+	}
+	apiObjs, err := c.g.GetCommits(ctx, git.GetCommitsArgs{
+		RepositoryId:   &repositoryId,
+		Project:        &projectId,
+		SearchCriteria: &searchCriteria,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Map the api object to our CommitType type
+	keys := make([]*commitType, 0, len(*apiObjs))
+	keys = append(keys, newCommit(c, &git.GitPush{
+		Commits: apiObjs,
+	}))
+
+	return keys, nil
+}
+
+func (c *CommitClient) Create(ctx context.Context, branch string, message string, files []gitprovider.CommitFile) (gitprovider.Commit, error) {
+	if len(files) == 0 {
+		return nil, fmt.Errorf("no files added")
+	}
+	repositoryId := c.ref.GetRepository()
+	projectId := c.ref.GetIdentity()
+	ref := "refs/heads/" + branch
+
+	var change []interface{}
+	var all []interface{}
+
+	for _, file := range files {
+		all = append([]interface{}{git.Change{
+			ChangeType: &git.VersionControlChangeTypeValues.Add,
+			Item: map[string]interface{}{
+				"path": file.Path,
+			},
+			NewContent: &git.ItemContent{
+				Content:     file.Content,
+				ContentType: &git.ItemContentTypeValues.RawText,
+			},
+			SourceServerItem: nil,
+			Url:              nil,
+		}}, change...)
+	}
+
+	commits := []git.GitCommitRef{
+		{
+			Changes: &all,
+			Comment: &message,
+		}}
+
+	// get latest commit from branch
+
+	commitsList, err := c.ListPage(ctx, branch, 0, 1)
```

Comment:
can we avoid this api call?  If we specify the reference in `git.GitRefUpdate` can it create the commit on top and update the reference without us telling it?

Reply:
yes this is possible, it would mean only using the `ref` and the pull request would still be created

Reply:
looks like i was looking at the wrong docs above, for creating a push

Reply:
i had a look and we could do something like this:-

 	`latestCommitTreeSHA := "0000000000000000000000000000000000000000"` although a behaviour i've noticed is that it would only work this way if it is in a new branch, otherwise to update the same branch we would need to get the latest commit.

    Another option is to support both of this, have a method for creating a new commit in a new branch +  a method to add a commit to an exisiting branch

Reply:
do you think we can open an issue for this? it feels weird that we have to do this. But in the meantime I guess we have to leave it as it is.

Reply:
yes, let me do that, yeah unfortunately

Reply:
https://developercommunity.visualstudio.com/t/Azure-REST-API---GIT-PUSH-providing-oldO/10289807 i'll keep an eye on it and as well follow up internally